### ISSUE TITLE
[Feature] title

### ISSUE THREAD

**Issue by firelizzard18** [gotk3/gotk3#787]
**Is your feature request related to a problem? Please describe.**
Given an app defined as follows:

```go
func main() {
	gtk.Init(&os.Args)

	app, err := gtk.ApplicationNew("com.example.app", glib.APPLICATION_HANDLES_OPEN)
	if err != nil {
		log.Fatal(err)
	}

	// ...
	os.Exit(app.Run(os.Args))
}
```

When this app is executed, GTK generates two signals. If I execute `./app`, the signals are `startup` and `activate`, in that order. If I execute `./app some-file`, the signals are `startup` and `open`. The expected signature of the callback for `open` is the same as `g_application_open (GApplication *application, GFile **files, gint n_files, const gchar *hint);`

**Describe the solution you'd like**
I would like to be able to `app.Connect("open", func(*gtk.Application, []*glib.File, string) { ... })`.

**Describe alternatives you've considered**
```go
func open(app *gtk.Application, filePtrs unsafe.Pointer, count int, hint string) {
	files := make([]string, count)
	for i := range files {
		ptr := (*unsafe.Pointer)(unsafe.Pointer(uintptr(filePtrs) + unsafe.Sizeof(uintptr(0))*uintptr(i)))
		f := &glib.File{
			Object: &glib.Object{
				GObject: glib.ToGObject(*ptr),
			},
		}
		files[i] = f.GetPath()
	}
	log.Print(files)
}
```

**Additional context**
As far as I can tell from debugging the Go parts*, glib/gtk passes `files` as a plain pointer GValue - that is, `v.Type()` on the `glib.Value` returns `TYPE_POINTER, TYPE_POINTER`. So, AFAIK, glib/gtk does not provide the necessary metadata out of the box to detect that this is an array of GFiles.

# Full example

* Screenshot of [`./app`](https://user-images.githubusercontent.com/879055/120938724-f1b05480-c6d9-11eb-97c2-97378753fca3.png)
* Screenshot of [`./app /tmp/foo /tmp/bar`](https://user-images.githubusercontent.com/879055/120938760-28866a80-c6da-11eb-9423-b9580eb07c89.png)

```go
package main

import (
	"log"
	"os"
	"unsafe"

	"github.com/gotk3/gotk3/glib"
	"github.com/gotk3/gotk3/gtk"
)

func main() {
	gtk.Init(&os.Args)

	app, err := gtk.ApplicationNew("com.example.app", glib.APPLICATION_HANDLES_OPEN)
	if err != nil {
		log.Fatal(err)
	}

	app.Connect("startup", startup)
	app.Connect("open", open)
	app.Connect("activate", activate)
	os.Exit(app.Run(os.Args))
}

func startup(app *gtk.Application) {
	win, err := gtk.ApplicationWindowNew(app)
	if err != nil {
		log.Fatal(err)
	}

	win.Connect("destroy", func() { app.Quit() })

	win.SetTitle("Hello!")
	win.SetDefaultSize(400, 300)
}

func activate(app *gtk.Application) {
	win := app.GetActiveWindow()

	l, err := gtk.LabelNew("Hello, gotk3!")
	if err != nil {
		log.Fatal("Unable to create label:", err)
	}

	win.Add(l)
	win.ShowAll()
}

func open(app *gtk.Application, filePtrs unsafe.Pointer, count int, hint string) {
	grid, err := gtk.GridNew()
	if err != nil {
		log.Fatal("Unable to create grid:", err)
	}

	app.GetActiveWindow().Add(grid)
	grid.SetOrientation(gtk.ORIENTATION_VERTICAL)

	l, err := gtk.LabelNew("You got files!")
	if err != nil {
		log.Fatal("Unable to create label:", err)
	}
	grid.Add(l)

	for i := 0; i < count; i++ {
		ptr := (*unsafe.Pointer)(unsafe.Pointer(uintptr(filePtrs) + unsafe.Sizeof(uintptr(0))*uintptr(i)))
		f := &glib.File{
			Object: &glib.Object{
				GObject: glib.ToGObject(*ptr),
			},
		}

		l, err := gtk.LabelNew(f.GetPath())
		if err != nil {
			log.Fatal("Unable to create label:", err)
		}
		grid.Add(l)
	}

	app.GetActiveWindow().ShowAll()
}
```

**Comment by ahsand97** [comment]
The callbacks of the function `Connect` have to respect the signal signature (meaning it should have the same parameters and return type defined in the signal), the signal  `open` is defined with this signature https://docs.gtk.org/gio/signal.Application.open.html

So the callback must have those parameters in order to successfully connect to the signal and react to its emition, you can create an intermmediate function to translate `files` and `count` to create a `[]*glib.File` and then call your desired callback using it.

**Comment by firelizzard18** [comment]
> The callbacks of the function Connect have to respect the signal signature

@ahsand97 This is not strictly true. The callbacks passed to `Connect` accept Go types, not C types. The method in the `glib` package that unmarshals the callback parameters already converts the C types to Go types. I am suggesting that method should be updated to allow the callback for `open` to accept `[]*glib.File` directly. Having dug through the code, I am confident that there is no technological limitation that forces the Go functions to have the exact same structure as the C functions.

**Comment by ahsand97** [comment]
> The callbacks passed to `Connect` accept Go types, not C types.

Yes, this is true , the callback should have the Go types representing the C types from the signal signature.

`Connect` is a wrapper around `signal_connect_closure`, all the closures are created with the [same](https://github.com/gotk3/gotk3/blob/1d544513fb748b660717266a1fe0284daeeb8fbc/glib/glib.go.h#L151) ClosureMarshal 

the ClosureMarshal `goMarshal` is defined [here](https://github.com/gotk3/gotk3/blob/1d544513fb748b660717266a1fe0284daeeb8fbc/glib/glib.go#L182) and basically takes every single parameter from C and converts them to valid Go type and then calls the callback with those types, meaning it just translates the params from C to Go.

The ClosureMarshal you propose is very case-specific, meaning it would only work with the signal `open`, and it would require a specific logic, something like checking for the params in positions:

1. `gpointer files`
2. `gint n_files`

and then create a `[]*glib.File` with the content of `files` and size of `n_files` to then call the callback with it.

Also, the `Connect` function would require to be modified and check if the param `detailedSignal` is equal to "open" to assign the specific ClosureMarshal, but again, it seems to be very case-specific.

I still think the easiest way is to create an intermediate function to translate the received params of the callback to your desired ones and then call the real callback function, or, you can use Cgo to create your own Closure, ClosureMarshal and then use `signal_connect_closure` to connect the desired signal with the closure but I think is much more complicated.
